proposals/NNNN-mixins.md

```diff
@@ -0,0 +1,499 @@
+      logger.log("Error", 3)
+    }
+    logger.log("Ending", 1)
+  }
+}
+```
+
+## Detailed design
+
+### Mixin members 
+
+Property requirements:
+
+```swift
+var name: Type { get set }
+```
+
+Property definitions:
+
+```swift
+var name: Type
+```
+
+Method requirements:
+
+```swift
+func name(arg: Type) -> Result
+```
+
+Method definitions:
+
+```swift
+func name(arg: Type) -> Result { }
+```
+
+There is no notion of default methods in mixins. Methods of mixins require `override` to redefine them in types and can be `final`. 
+
+Mixins consider themselves as value types, as protocols and structs. Every member and requirement, which needs to mutate `self`, needs to be declared as `mutating`.
+
+Subscript requirements and definitions.
+
+Initializer requirements:
+
+```swift
+init(name: Type)
+```
+
+Initializer definitions:
+
+```swift
+init(name: Type) { }
+```
+
+Properties can be immediately initialized: `var name: Type = initializer`
+
+These initialized properties are prepended to all initializers. If no initializers are defined, an implicit default initializer is created.
+
+### Initializers
+
+Mixins can have implemented initializers. They act like initializers of classes. `super` initializer calls are prefixed with parent protocol name for disambiguation purposes. `super` initializer must be explicitly called in initializer of subtype. The syntax for that is `SuperMixinName.super.init()`:
+
+```swift
+mixin A {
+  init(newPath: String) { }
+}
+
+mixin B: A {
+  override init(newPath: String) { }
+}
+```
+
+Default `super` initializers also need to be called explicitly. In multiple inheritance, it is a responsibility of subtype to ensure that initialization order matches semantics of mixins it mixes in:
+
+```swift
+mixin A { }
+mixin B { }
+
+mixin C: A, B {
+  init() {
+    A.super.init()
+    B.super.init()
+  }
+}
+```
+
+Note that `super` syntax can also be used for calling supertype members, as in classes.
+
+### Mixin inheritance
+
+If a single mixin is mixed-in using different paths, then only one copy of it will be kept in the final struct/class. Example:
+
+```swift
+mixin A { var x: Int = 1 }
+mixin B: A { }
+mixin C: A { }
+struct D: B, C { }  // contains a single copy of x
+```
+
+It works as if bodies of all mixins indirectly mixed-in into `D` were merged directly into `D`:
+
+```swift
+mixin ASelf { var x: Int = 1 }
+mixin BSelf { }
+mixin CSelf { }
+struct D: ASelf, BSelf, CSelf { }
+```
+
+Members with conflicting signatures inherited from two different places result in a compilation error.
+
+```swift
+mixin A { var x: Int = 0 }
+mixin B { var x: Int = 1 }
+struct C : A, B { }  // error
+```
+
+### Resolving conflicts
+
+We can resolve conflicts by renaming one or both of the conflicting members. Syntax for renaming conflicting members is:
+
+```swift
+var y = B.x
+func g(arg: Type) = B.f
+```
+
+Example:
+
+```swift
+mixin A { var x: Int = 0 }
+mixin B { var x: Int = 1 }
+struct C : A, B { var y = B.x }
```

Comment:
I think this is a typo? This should still be a compilation error (it's not clear whether it's using A.x or B.x), unless the proposal is `var x = B.x` (in which case this a typo that needs to be updated).